File: multi_language_code_generator/implementations/java/DependencyResolver.java
import java.util.*;

public class DependencyResolver {

    // Simple DTO for the result to match JSON structure
    public static class ResolutionResult {
        public boolean success;
        public List<String> build_order;
        public String error;

        public ResolutionResult(boolean success, List<String> build_order, String error) {
            this.success = success;
            this.build_order = build_order;
            this.error = error;
        }
    }

    public ResolutionResult resolve(Map<String, List<String>> dependencies) {
        Map<String, Integer> inDegree = new HashMap<>();
        Map<String, List<String>> graph = new HashMap<>();
        Set<String> allNodes = new HashSet<>(dependencies.keySet());

        // 1. Normalize and Identify all nodes
        for (List<String> deps : dependencies.values()) {
            allNodes.addAll(deps);
        }

        // Initialize in-degree for all nodes to 0
        for (String node : allNodes) {
            inDegree.put(node, 0);
            graph.put(node, new ArrayList<>());
        }

        // 2. Build Graph (Edge: Dependency -> Dependent)
        // If A depends on B (A -> B), we need B before A.
        // So in our graph for topo sort: B -> A.
        for (Map.Entry<String, List<String>> entry : dependencies.entrySet()) {
            String pkg = entry.getKey();
            List<String> deps = entry.getValue();
            
            for (String dep : deps) {
                graph.get(dep).add(pkg);
                inDegree.put(pkg, inDegree.get(pkg) + 1);
            }
        }

        // 3. Initialize Queue with 0 in-degree nodes
        // PriorityQueue used to ensure deterministic output (lexicographical order)
        PriorityQueue<String> queue = new PriorityQueue<>();
        for (String node : allNodes) {
            if (inDegree.get(node) == 0) {
                queue.add(node);
            }
        }

        List<String> sortedOrder = new ArrayList<>();

        // 4. Process Queue
        while (!queue.isEmpty()) {
            String current = queue.poll();
            sortedOrder.add(current);

            List<String> neighbors = graph.get(current);
            // Sort neighbors strictly for deterministic output across languages
            Collections.sort(neighbors); 
            
            for (String neighbor : neighbors) {
                inDegree.put(neighbor, inDegree.get(neighbor) - 1);
                if (inDegree.get(neighbor) == 0) {
                    queue.add(neighbor);
                }
            }
        }

        // 5. Check for Cycles
        if (sortedOrder.size() != allNodes.size()) {
            return new ResolutionResult(false, null, "Cycle detected or unresolvable dependencies.");
        }

        return new ResolutionResult(true, sortedOrder, null);
    }
}

File: multi_language_code_generator/implementations/python/dependency_resolver.py
from collections import deque, defaultdict
import json

class DependencyResolver:
    def resolve(self, dependencies: dict) -> dict:
        """
        Resolves build order using Kahn's Algorithm.
        """
        in_degree = defaultdict(int)
        graph = defaultdict(list)
        all_nodes = set(dependencies.keys())

        # Normalize graph: Ensure all referenced nodes exist in set
        for pkg, deps in dependencies.items():
            for dep in deps:
                all_nodes.add(dep)

        # Build graph: dependent -> dependency (reversed logic for build order)
        # However, for "install dependencies first", if A depends on B:
        # We need B before A. Edge: B -> A.
        for pkg, deps in dependencies.items():
            for dep in deps:
                graph[dep].append(pkg)
                in_degree[pkg] += 1
        
        # Initialize queue with nodes having 0 dependencies
        queue = deque([node for node in all_nodes if in_degree[node] == 0])
        
        # Sort initial queue for deterministic output in tests
        # Note: deque doesn't support sort directly, so we sort a list then make a deque
        sorted_start = sorted(list(queue))
        queue = deque(sorted_start)
        
        sorted_order = []

        while queue:
            current = queue.popleft()
            sorted_order.append(current)

            if current in graph:
                # Sort neighbors for deterministic behavior
                neighbors = sorted(graph[current])
                for neighbor in neighbors:
                    in_degree[neighbor] -= 1
                    if in_degree[neighbor] == 0:
                        queue.append(neighbor)
        
        if len(sorted_order) != len(all_nodes):
            return {
                "success": False,
                "build_order": None,
                "error": "Cycle detected or unresolvable dependencies."
            }

        return {
            "success": True,
            "build_order": sorted_order,
            "error": None
        }
